feat: add Toc component

tonai opened the pull request:
Adding a built-in TOC component.

TOC is really useful when you have long presentations (like for a 3 days training for example). It helps people not to be to much lost.
Lots of presentation library miss this feature, that is mandatory in the usage I have.
I found this library awesome, that's why I am proposing this feature, because It would even more incredible with this feature :smiley: 

This feature automatically gather informations about the title and the title level of each slides to build a tree.
You can also set both information using front matter syntax (if your slide does not contain a title, or if you want to override it).
For example:
```md
---
title: Amazing slide title
title level: 2
---
```

Available props for the Toc component:
* columns (default = 1): TOC in multiple columns
* maxDepth (default = Infinity): Maximum depth of the TOC
* minDepth (default = 1): Minimum depth of the TOC
* mode (default = 'all'):
  * 'all': Displays all items
  * 'onlyCurrentTree': Displays only items that belong to the current active slide (active slide, parents of active slides and children of active slide)
  * 'onlySiblings': Displays only items that belong to the current active slide and their siblings (but it will not show the children of a slide that does not belong to the current tree)

You can also mark a slide to not be displayed in the TOC with the front matter syntax:
```md
---
skipInToc: true
---
```

Examples (Table of content slide is skipped):

=> `columns="3"`
![screencapture-localhost-3031-2-2022-01-12-16_59_50](https://user-images.githubusercontent.com/5246045/149179362-679f42f6-eda5-41b2-a951-3ac77c350b2e.png)

=> `maxDepth="2"`
![screencapture-localhost-3030-12-2022-01-12-17_03_38](https://user-images.githubusercontent.com/5246045/149179576-a5a35b98-9d86-40ea-ba08-8835ef6ec5eb.png)

=> `minDepth="2"`
![screencapture-localhost-3030-21-2022-01-12-17_04_04](https://user-images.githubusercontent.com/5246045/149179625-72767700-a39c-4079-b830-baa416a2afb8.png)

=> `mode="onlyCurrentTree"`
![screencapture-localhost-3031-17-2022-01-12-17_00_36](https://user-images.githubusercontent.com/5246045/149179849-cea41395-614d-42b6-b553-3e3d2a2058f2.png)

=> `mode="onlySiblings"`
![screencapture-localhost-3031-31-2022-01-12-17_01_15](https://user-images.githubusercontent.com/5246045/149179926-f5b8d630-3a98-49dd-9bf6-a827bc2b467a.png)

Possible enhancements:
* add some styles to see active item in the list (classes are already added but I am not sure where I should put the styles or if we should let the user add his own styles)
* the tree for the TOC is a computed value inside the Toc component, but if you put multiple Toc component in your presentation, the tree get calculated for each components. Maybe we can calculate the tree globally and get it inside the component (that will filter it accordingly to the props), but I am not sure how we should do that ().

antfu commented on packages/client/builtin/Toc.vue:
```diff
@@ -0,0 +1,107 @@
+
+<Toc columns="2" maxDepth="3" mode="onlySiblings"/>
+-->
+<script lang="ts">
```
Could we use the `<script setup>` syntax for overall consistency?

tonai replied:
Of course no problem, but I'm not completely familiar with this new syntax and I don't know how I can access the $slidev inside of it. Can you give me an hint ?

antfu replied:
`getCurrentInstance().proxy.$slidev` I think. But actually you can directly import them from the internal modules

tonai replied:
Thank you for your answer.
I have pushed the update.
Also what is your thought about the last 2 points ?
* add some styles to see active item in the list (classes are already added but I am not sure where I should put the styles or if we should let the user add his own styles)
* the tree for the TOC is a computed value inside the Toc component, but if you put multiple Toc component in your presentation, the tree get calculated for each components. Maybe we can calculate the tree globally and get it inside the component (that will filter it accordingly to the props), but I am not sure how we should do that.

For the last point can I add a computed value for the tree inside the `@slidev/client/logic/nav` ? Will it solves the problem ?

antfu replied:
1. I prefer to not have styles, but we could still add classes prefixed by `.slidev` just like other classes for users to override for certain state
2. I don't mind that as I don't think the computed will be very expensive. We could leave it as an future optimization if it really leads to performance issues

antfu commented on packages/client/builtin/Toc.vue:
```diff
@@ -0,0 +1,122 @@
+      children: [],
+      level,
+      path: route.path,
+      skipInToc: Boolean(route.meta?.skipInToc),
```
How about `hideInToc`?

antfu commented on packages/client/routes.ts:
```diff
@@ -51,6 +51,8 @@ declare module 'vue-router' {
       no: number
       filepath: string
+      title?: string
+      titleLevel?: number
```
How about just `level`?

antfu reviewed (commented):
Looks great, thanks!

tonai commented:
Thanks for your comments, I fixed them both.
I also moved the computed tree value inside the nav logic file, not really for performance issue, but I think the tree can be reused in ome situation (like for creating a sidebar reported in this issue: https://github.com/slidevjs/slidev/discussions/439)

antfu commented:
@tonai would be great if you can also help adding this to the docs, thanks!

antfu commented:
You can add a sub section of something like `### TOC`, and I will fulfill the rest comoponents :)
